Mathlib/FieldTheory/JacobsonNoether.lean

```diff
@@ -0,0 +1,198 @@
+-/
+
+namespace JacobsonNoether
+
+variable {D : Type*} [DivisionRing D] [Algebra.IsAlgebraic (Subring.center D) D]
+
+local notation3 "k" => Subring.center D
+
+open Polynomial LinearMap LieAlgebra
+
+/-- If `D` is a purely inseparable extension of `k` with characteristic `p`,
+  then for every element `a` of `D`, there exists a natural number `n`
+  such that `a ^ (p ^ n)` is contained in `k`. -/
+lemma exists_pow_mem_center_of_inseparable (p : ℕ) [hchar : ExpChar D p] (a : D)
+    (hinsep : ∀ x : D, IsSeparable k x → x ∈ k) : ∃ n, a ^ (p ^ n) ∈ k := by
+  have := (@isPurelyInseparable_iff_pow_mem k D _ _ _ _ p (ExpChar.expChar_center_iff.2 hchar)).1
+  have pure : IsPurelyInseparable k D := ⟨Algebra.IsAlgebraic.isIntegral, fun x hx ↦ by
+    rw [RingHom.mem_range, Subtype.exists]
+    exact ⟨x, ⟨hinsep x hx, rfl⟩⟩⟩
+  obtain ⟨n, ⟨m, hm⟩⟩ := this pure a
+  have := Subalgebra.range_subset (R := k) ⟨(k).toSubsemiring, fun r ↦ r.2⟩
+  exact ⟨n, Set.mem_of_subset_of_mem this <| Set.mem_range.2 ⟨m, hm⟩⟩
+
+/-- If `D` is a purely inseparable extension of `k` with characteristic `p`,
+  then for every element `a` of `D \ k`, there exists a natural number `n`
+  **greater than 0** such that `a ^ (p ^ n)` is contained in `k`. -/
+lemma exists_pow_mem_center_of_inseparable' (p : ℕ) [ExpChar D p] {a : D}
+    (ha : a ∉ k) (hinsep : ∀ x : D, IsSeparable k x → x ∈ k) : ∃ n, 1 ≤ n ∧ a ^ (p ^ n) ∈ k := by
+  obtain ⟨n, hn⟩ := exists_pow_mem_center_of_inseparable p a hinsep
+  by_cases nzero : n = 0
+  · rw [nzero, pow_zero, pow_one] at hn
+    exact (ha hn).elim
+  · exact ⟨n, ⟨Nat.one_le_iff_ne_zero.mpr nzero, hn⟩⟩
+
+/-- If `D` is a purely inseparable extension of `k` of characteristic `p`,
+  then for every element `a` of `D \ k`, there exists a natural number `m`
+  greater than 0 such that `(a * x - x * a) ^ n = 0` (as linear maps) for
+  every `n` greater than `(p ^ m)`. -/
+lemma exist_pow_eq_zero_of_le (p : ℕ) [hchar : ExpChar D p]
+    {a : D} (ha : a ∉ k) (hinsep : ∀ x : D, IsSeparable k x → x ∈ k):
+  ∃ m, 1 ≤ m ∧ ∀ n, p ^ m ≤ n → (ad k D a)^[n] = 0 := by
+  obtain ⟨m, hm⟩ := exists_pow_mem_center_of_inseparable' p ha hinsep
+  refine ⟨m, ⟨hm.1, fun n hn ↦ ?_⟩⟩
+  have inter : (ad k D a)^[p ^ m] = 0 := by
+    ext x
+    rw [ad_eq_lmul_left_sub_lmul_right, ← pow_apply, Pi.sub_apply,
+      sub_pow_expChar_pow_of_commute p m (commute_mulLeft_right a a), sub_apply,
+      pow_mulLeft, mulLeft_apply, pow_mulRight, mulRight_apply, Pi.zero_apply,
+      Subring.mem_center_iff.1 hm.2 x]
+    exact sub_eq_zero_of_eq rfl
+  rw [(Nat.sub_eq_iff_eq_add hn).1 rfl, Function.iterate_add, inter, Pi.comp_zero,
+    iterate_map_zero, Function.const_zero]
+
+variable (D) in
+/-- Jacobson-Noether theorem: For a non-commutative division algebra
+  `D` that is algebraic over its center `k`, there exists an element
+  `x` of `D \ k` that is separable over `k`. -/
+theorem exists_separable_mem_of_not_central (H : k ≠ (⊤ : Subring D)) :
```

Comment:
I would think `of_not_comm` is more fitting. Similarly below, the assumptions say something more than a prime.